isisd/isis_sr.c

```diff
@@ -438,7 +438,7 @@ void isis_sr_prefix_cfg2subtlv(const struct sr_prefix_cfg *pcfg, bool external,
 	}
 	if (external)
 		SET_FLAG(psid->flags, ISIS_PREFIX_SID_READVERTISED);
-	if (pcfg->node_sid)
+	if (pcfg->node_sid && !pcfg->n_flag_clear)
```

Comment:
This is indeed necessary to handle runtime changes of the "n-flag-clear" configuration option. I wonder if we can remove the other `pcfg->n_flag_clear` checks in the code since this one alone might catch all possible cases.

Reply:
Humm. I can look into that, but will need some time.

Reply:
@rwestphal I've looked into your question.

The check of pcfg->n_flag_clear appears in very few places. Esssentially:
* to read the value (nb_config). This is needed.
* when configuring a new prefix SID. This is also needed.
* In the translation function to TLVs. This is also needed.
* In the if_new_hook. The comment there is that it is neeeded since isisd reads the config before learning the interfaces (I guess those are needed to determine if they are loopback or not).

So, the way the code is, I doubt it pays off to remove any other check.
If anything, the node_sid variable appears mostly in the same places. Was your question actually merging the two variables into one? I guess that could be done, but fear it goes against readability? To be node_sid is state and n_flag_clear configuration. I would not mix both.

Reply:
One last bit regarding the checks of the n_flag_clear. I would say that the only place where it could be removed is the sr_if_new_hook() function, given that the (un)setting of the real flag is in isis_sr_prefix_cfg2subtlv() which may get called as a result of a lsp_regenerate_schedule(). However, by removing it there, we would be regenerating the LSP unnecessarily?

Reply:
ci:rerun

Reply:
> The check of pcfg->n_flag_clear appears in very few places. Esssentially:
> 
> * to read the value (nb_config). This is needed.
> * when configuring a new prefix SID. This is also needed.
> * In the translation function to TLVs. This is also needed.
> * In the if_new_hook. The comment there is that it is neeeded since isisd reads the config before learning the interfaces (I guess those are needed to determine if they are loopback or not).

My rationale is the following: in the second and fourth cases, `lsp_regenerate_schedule()` is called after the existing `pcfg->n_flag_clear` checks, which renders them redundant since `isis_sr_prefix_cfg2subtlv()` will be called for all configured Prefix-SIDs (setting `ISIS_PREFIX_SID_NODE` or not according to the configuration). So the idea is to simplify the code by having the `pcfg->n_flag_clear` check performed only once (one check to rule them all!).

Reply:
In the case of the `sr_if_new_hook()` function, I think it's better to always set `pcfg->node_sid` to `true` for loopback routes regardless if "n-flag-clear" is configured or not, since the Node-SID information might be exposed in the future though YANG for example.

Reply:
@rwestphal I've removed the redundant checks in a second patch. Please, let me know if you're fine with the changes.

Reply:
@Fredi-raspall thanks for the update, looks great to me!

Reply:
ok. Thanks @rwestphal !